framework/test/gtest/BundleTrackerConcurrencyTest.cpp

```diff
@@ -0,0 +1,232 @@
+#include "cppmicroservices/Bundle.h"
+#include "cppmicroservices/BundleTracker.h"
+#include "cppmicroservices/Framework.h"
+#include "cppmicroservices/FrameworkEvent.h"
+#include "cppmicroservices/FrameworkFactory.h"
+
+#include <future>
+
+#include "TestUtils.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using namespace cppmicroservices;
+
+class BundleTrackerConcurrencyTest : public ::testing::Test
+{
+protected:
+  Framework framework;
+  BundleContext context;
+  static constexpr BundleTracker<>::BundleStateMaskType all_states =
+    BundleTracker<>::CreateStateMask(Bundle::State::STATE_ACTIVE,
+                                     Bundle::State::STATE_INSTALLED,
+                                     Bundle::State::STATE_RESOLVED,
+                                     Bundle::State::STATE_STARTING,
+                                     Bundle::State::STATE_STOPPING,
+                                     Bundle::State::STATE_UNINSTALLED);
+
+public:
+  BundleTrackerConcurrencyTest()
+    : framework(FrameworkFactory().NewFramework()){};
+
+  ~BundleTrackerConcurrencyTest() override = default;
+
+  void SetUp() override
+  {
+    framework.Start();
+    context = framework.GetBundleContext();
+  }
+
+  void TearDown() override
+  {
+    framework.Stop();
+    framework.WaitForStop(std::chrono::milliseconds::zero());
+  }
+};
+
+MATCHER_P(HasName, name, "")
+{
+  return arg.GetSymbolicName() == name;
+}
+
+#ifdef US_ENABLE_THREADING_SUPPORT
+class MockCustomizer2 : public BundleTrackerCustomizer<>
+{
+public:
+  MOCK_METHOD(std::optional<Bundle>,
+              AddingBundle,
+              (const Bundle&, const BundleEvent&),
+              (override));
+  MOCK_METHOD(void,
+              ModifiedBundle,
+              (const Bundle&, const BundleEvent&, const Bundle&),
+              (override));
+  MOCK_METHOD(void,
+              RemovedBundle,
+              (const Bundle&, const BundleEvent&, const Bundle&),
+              (override));
+};
+
+TEST_F(BundleTrackerConcurrencyTest, ConcurrentOpenCloseWorks)
+{
+  auto bundleTracker = std::make_unique<BundleTracker<>>(context, all_states);
+
+  size_t numThreads = std::thread::hardware_concurrency();
+  ASSERT_GT(numThreads, 0ull) << "number of threads is 0";
+  std::vector<std::future<void>> futures;
+  std::promise<void> gate;
+  auto gateFuture = gate.get_future().share();
+  for (size_t i = 0; i <= numThreads; ++i) {
+    futures.push_back(
+      std::async(std::launch::async, [i, &bundleTracker, &gateFuture]() {
+        gateFuture.get();
+        for (int n = 0; n < 1000; ++n) {
+          if (i % 2 == 0) {
+            bundleTracker->Open();
+          } else {
+            bundleTracker->Close();
+          }
+        }
+      }));
+  }
+
+  gate.set_value();
+
+  for (auto& asyncFuture : futures) {
+    asyncFuture.get();
+  }
+}
+
+TEST_F(BundleTrackerConcurrencyTest, OpeningTrackerWhileBundlesChangeWorks)
```

Comment:
minor; rename testpoint name - remove "Works" and prepend "Test" to be more consistent with other test point names

Reply:
Done.